### DISCUSSION TITLE
Labelling and referencing a Matplotlib figure

### DISCUSSION THREAD

**Discussion by spring-haru** [executablebooks/meta#359] [Q&A]
I wish to create a label to a Matplotlib figure created inside a jb, and to refer to it somewhere else and jump to the figure (hyperlink). How do I do it?

**Comment by choldgraf** [comment]
Have you tried the instructions here?

https://jupyterbook.org/content/code-outputs.html#images

**Reply by spring-haru** [reply]
My understanding is that the instruction of the webpage is for an image created in advance. Please correct me if I am wrong. What I have in mind is a Matplotlib figure generated by code IN a Jupyter Notebook, i.e. a figure generated as an output shown just below a code block.

So, my question is if it is possible to label
(i) a  code block in general (which contains Matplotlib code in my case)
(ii)  a figure generated by (i)
(iii) a word/phase/sentence

If (iii) is possible, I just reference a word/phase/sentence near the code block or the figure to jump to it.

I am looking for something similar to the one explained at the following link:

https://jupyterbook.org/content/references.html#reference-section-labels

**Reply by choldgraf** [reply]
I think your best bet would be to use a section label that is close to the figure, and reference that.

For your question `(ii)`, does it not work if you use the same basic approach but with a Matplotlib figure instead of an image?

Could you also try using the "Glue" functionality: https://jupyterbook.org/content/code-outputs.html#the-glue-figure-directive ?

**Reply by martinlarsalbert** [reply]
@choldgraf how do you mean that you should do this more specifically?
I've been searching for an easy way to put numref and caption to an figure output from notebooks (typically a matplot figure), but so far I haven't found it. Does it mean that it does not exist or have a missed something obvious?

Normally I use nbconvert with my own preprocessors that takes the label and caption for the figures from the cell meta data. It would be so nice if jupyter book had something similiar! This is a deal breaker for me, as this is something I need when writing scientific papers.

**Reply by jangenoe** [reply]
I also like to have both the imported figures and the matplotlib generated figures numbered in the same, consistent way all over the jupyter book. At present, I use a dummy, empty figure of 1 pixel height (mpl.svg) below each matplotlib generated figure to add the md figure caption that I can reference throughout the book. For figures, it is a hack that works, but a consistent solution in the jb/my-st code would be preferred in the long run.

However, the same holds for the table caption. I can number the md table captions, but I would love to have the have the same consistent numbered table captions in the code generated tables. But whenever I generate an empty table in md, with a table caption, to reference the table from the markdown, the table is removed from the jb. I have not found a hack to solve this yet.
